src/structure/bitindex.rs

```diff
@@ -175,19 +214,41 @@ impl BitIndex {
         None
     }
 
-    pub fn select1_from_range(&self, subrank: u64, start: u64, end: u64) -> Option<u64> {
-        // todo this is a dumb implementation. we can actually do a much faster select by making sblock/block lookup ranged. for now this will work.
-        let rank_offset = if start == 0 { 0 } else { self.rank1(start - 1) };
+    fn select_from_range_opt_linear(
+        &self,
+        mut subrank: u64,
+        start: u64,
+        end: Option<u64>,
+        find: bool,
+    ) -> Option<u64> {
+        if end.is_some() && start >= end.unwrap() {
+            return None;
+        }
+
+        if subrank == 0 {
+            if self.array.get(start as usize) == find {
```

Comment:
```suggestion
            if self.get(start) == find {
```

Reply:
In addition, various other places could have used `.get()` or `.len()` direcltyo n the bitindex. I've now changed them all.